provisioner/provision.js

```diff
@@ -0,0 +1,302 @@
+    that.__watchdog = setTimeout(function() {
+      generalDebug('KILLING PROVISIONER BECAUSE IT APPEARS TO BE STUCK...');
+      // Hmm, should I instead just process.exit(1);
+      throw new Error('PROVISIONER HAS FALLEN AND CAN\'T GET BACK UP');
+    }, MAX_PROVISION_ITERATION);
+
+    var p = that.runAllProvisionersOnce();
+
+    p = p.then(function() {
+      if (that.__keepRunning && !process.env.PROVISION_ONCE) {
+        generalDebug('Done! Scheduling another provisioning iteration');
+        setTimeout(provisionIteration, that.provisionIterationInterval);
+      } else {
+        generalDebug('Done! Not scheduling another provisioning iteration');
+      }
+    });
+
+    p = p.catch(function(err) {
+      generalDebug('Error running a provisioning iteration');
+      console.error(err, err.stack);
+    });
+  }
+
+  provisionIteration();
+
+};
+
+
+/**
+ * Stop launching new provisioner iterations but don't
+ * end the current one
+ */
+Provisioner.prototype.stop = function () {
+  this.__keepRunning = false;
+  if (this.__watchdog) {
+    clearTimeout(this.__watchdog);
+  }
+};
+
+
+/**
+ * Run provisioners for all known worker types once
+ */
+Provisioner.prototype.runAllProvisionersOnce = function() {
+
+  var that = this;
+  // So that we get per-iteration strings
+  var debug = _debug(baseDbgStr + ':all:run_' + ++this.__provRunId);
+
+  debug('%s Beginning provisioning iteration', this.provisionerId);
+  var p = Promise.all([
+    this.WorkerType.loadAll(),
+    awsState(this.ec2, this.awsKeyPrefix),
+    // Remember that we cache pricing data!
+    this.pricingCache.get(),
+  ]);
+
+  p = p.then(function(res) {
+    var workerTypes = res[0];
+    var state = res[1];
+    var pricing = res[2];
+
+    that.reconcileNewlyTracked(debug, state);
+
+    debug('AWS knows of these workerTypes: %s', JSON.stringify(state.knownWorkerTypes()));
+    // We could probably combine this with the .map of workerTypes below... meh...
+    debug('There are workerType definitions for these: %s', JSON.stringify(workerTypes.map(function(x) {
+      return x.workerType;
+    })));
+
+    return Promise.all(workerTypes.map(function(workerType) {
+      // We should be able to filter by a specific workerType
+      var wtDebug = 
+        _debug(baseDbgStr + ':workerType_' + workerType.workerType + ':run_' + that.__provRunId);
+      return that.provisionType(wtDebug, workerType, state, pricing);
+    }));
+  });
+
+  p = p.then(function(res) {
+    debug('Completed provisioning iteration');
+    return res;
+  });
+
+  return p;
+};
+
+
+
+/**
+ * Provision a specific workerType.  This promise will have a value of true if
+ * everything worked.  Another option is resolving to the name of the worker to
+ * make it easier to see which failed, but I'd prefer that to be tracked in the
+ * caller. Note that awsState as passed in should be specific to a workerType
+ */
+Provisioner.prototype.provisionType = function(debug, workerType, state, pricing) {
```

Comment:
For now, I'd like to leave the pricing as it is.  I moved creating the `AwsManager` into the bin/provisioner script and am now injecting it into the `Provisioner`.  This makes it way easier to test the Provisioning logic should we want to.